File: src/Roguelike/Core/Combat/Combatant/Hero.cs
using Roguelike.Data;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Roguelike.Core
{
    /// <summary>
    /// Represents the player's character, holding all their run-time state
    /// </summary>
    public class Hero : Combatant
    {
        public HeroData SourceHeroData => (HeroData)SourceData;

        public DeckManager Deck { get; }
        public int CurrentMana { get; set; }
        public int MaxMana { get; set; }
        public int CurrentGold { get; set; }
        public List<RelicData> Relics { get; } = new List<RelicData>();

        public Hero(HeroData sourceData, Random rng) : base(sourceData)
        {
            Deck = new DeckManager(rng);
            MaxMana = sourceData.StartingMana;
            CurrentMana = MaxMana;
            CurrentGold = sourceData.StartingGold;
        }

        public void StartTurn()
        {
            Block = 0;
            CurrentMana = MaxMana;

            var philosophicalEffect = ActiveEffects.FirstOrDefault(e => 
                e.SourceData is StatusEffectData s && s.EffectType == StatusEffectType.Philosophical);
            
            if (philosophicalEffect != null)
            {
                var philData = (StatusEffectData)philosophicalEffect.SourceData;
                CurrentMana += philData.Intensity;
            }

            Deck.DrawCards(SourceHeroData.StartingHandSize);
        }
    }
}

File: python_roguelike/core/hero.py
import random
from typing import List, Optional
from .combatant import Combatant
from .deck_manager import DeckManager
from .active_effect import ActiveEffect
from ..data.combatant.hero_data import HeroData
from ..data.relic_data import RelicData
from ..data.status_effect_data import StatusEffectData
from ..data.enums import StatusEffectType


class Hero(Combatant):
    def __init__(self, source_data: HeroData, rng: random.Random):
        super().__init__(source_data)
        self.source_hero_data: HeroData = source_data
        self.deck: DeckManager = DeckManager(rng)
        self.max_mana: int = source_data.starting_mana
        self.current_mana: int = self.max_mana
        self.current_gold: int = source_data.starting_gold
        self.relics: List[RelicData] = []

    def start_turn(self):
        self.block = 0
        self.current_mana = self.max_mana

        for ae in self.active_effects:
            if (isinstance(ae.source_data, StatusEffectData) and
                    ae.source_data.effect_type == StatusEffectType.Philosophical):
                self.current_mana += ae.source_data.intensity
                break

        self.deck.draw_cards(self.source_hero_data.starting_hand_size)
